File: src/main/java/ash/java/graphql/fields/MultiSearchSchema.java
package ash.java.graphql.fields;

import ash.java.graphql.data.SearchDao;
import ash.java.graphql.types.movie.MovieType;
import ash.java.graphql.types.multisearch.PersonType;
import ash.java.graphql.types.multisearch.TvShowType;
import graphql.schema.GraphQLFieldDefinition;
import graphql.schema.GraphQLList;
import graphql.schema.GraphQLNonNull;
import graphql.schema.GraphQLUnionType;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import static graphql.Scalars.*;
import static graphql.schema.GraphQLUnionType.newUnionType;
import static graphql.schema.GraphQLFieldDefinition.newFieldDefinition;

@Service
public class MultiSearchSchema implements FieldProducer {

    private SearchDao searchDao;

    @Autowired
    public MultiSearchSchema(SearchDao searchDao) {
        this.searchDao = searchDao;
    }

    private GraphQLUnionType multiSearchType = newUnionType()
            .name("MultiSearchType")
            .possibleType(new PersonType().getGraphQlType())
            .possibleType(new MovieType().getGraphQlType())
            .possibleType(new TvShowType().getGraphQlType())
            .typeResolver(object -> {
                if (object instanceof PersonType) {
                    return new PersonType().getGraphQlType();
                }
                if (object instanceof MovieType) {
                    return new MovieType().getGraphQlType();
                }
                if (object instanceof TvShowType) {
                    return new TvShowType().getGraphQlType();
                }
                return null;
            })
            .build();

    private GraphQLFieldDefinition multiSearchField = newFieldDefinition()
            .name("multiSearch")
            .type(new GraphQLList(multiSearchType))
            .argument(arg -> arg.name("query").type(new GraphQLNonNull(GraphQLString)))
            .argument(arg -> arg.name("language").type(GraphQLString))
            .argument(arg -> arg.name("page").type(GraphQLInt))
            .argument(arg -> arg.name("include_adult").type(GraphQLBoolean))
            .argument(arg -> arg.name("region").type(GraphQLString))
            .dataFetcher(env -> searchDao.searchMultiSearch(env.getArguments()))
            .build();

    @Override
    public GraphQLFieldDefinition getFieldDefinition() {
        return multiSearchField;
    }
}

File: src/main/kotlin/ash/kotlin/graphql/fields/MultiSearchSchema.kt
package ash.kotlin.graphql.fields

import ash.kotlin.graphql.data.TmdbUtil
import ash.kotlin.graphql.types.movie.MovieType
import ash.kotlin.graphql.types.multisearch.PersonType
import ash.kotlin.graphql.types.multisearch.TvShowType
import graphql.Scalars.*
import graphql.schema.GraphQLFieldDefinition
import graphql.schema.GraphQLList
import graphql.schema.GraphQLNonNull
import graphql.schema.GraphQLUnionType
import javax.inject.Inject

class MultiSearchSchema @Inject constructor(private val tmdbUtil: TmdbUtil) : FieldDefiner
{
    override fun getFieldDefinition(): GraphQLFieldDefinition
    {
        return createMultiSearchField()
    }

    private fun createMultisearchUnionType(): GraphQLUnionType
    {
        return GraphQLUnionType.newUnionType()
                .name("MultiSearchType")
                .possibleType(PersonType().getGraphQlType())
                .possibleType(MovieType().getGraphQlType())
                .possibleType(TvShowType().getGraphQlType())
                .typeResolver {
                    when (it)
                    {
                        is PersonType -> PersonType().getGraphQlType()
                        is MovieType -> MovieType().getGraphQlType()
                        is TvShowType -> TvShowType().getGraphQlType()
                        else -> null
                    }
                }
                .build()
    }

    private fun createMultiSearchField(): GraphQLFieldDefinition
    {
        return GraphQLFieldDefinition.newFieldDefinition()
                .name("multiSearch")
                .type(GraphQLList(createMultisearchUnionType()))
                .argument { arg -> arg.name("query").type(GraphQLNonNull(GraphQLString)) }
                .argument { arg -> arg.name("language").type(GraphQLString) }
                .argument { arg -> arg.name("page").type(GraphQLInt) }
                .argument { arg -> arg.name("include_adult").type(GraphQLBoolean) }
                .argument { arg -> arg.name("region").type(GraphQLString) }
                .dataFetcher { env -> tmdbUtil.searchMulti(env.arguments.toList()) }
                .build()
    }
}